### PR TITLE
Cross platform ActivityIndicator

### PR DESCRIPTION
**Pull request by janicduplessis**
The API for `ActivityIndiatorIOS` and `ProgressBarAndroid` is very similar and can be merged in a cross platform component that displays a circular indeterminate loading indicator.

This deprecates `ActivityIndiatorIOS` and non-horizontal `ProgressBarAndroid` in favor of this new component.

**Test plan (required)**

Tested with the ActivityIndicator example in UIExplorer on android and ios. Also made sure that `ActivityIndicatorIOS` still works and displays a deprecation warning. Also tested that `ProgressBarAndroid` with `indeterminate == true` and `styleAttr != 'Horizontal'` displays a deprecation warning.

### CODE REVIEW

**Comment by mkonicek** [comment]
cc @dmmiller

**Comment by mkonicek** [comment]
Very cool! Have you considered naming this `Spinner`? I hear people often refer to this component as spinner.

**Comment by janicduplessis** [comment]
I think Spinner could be confusing with http://developer.android.com/guide/topics/ui/controls/spinner.html for Android devs and since we already had ActivityIndicatorIOS I went with the iOS name.

**Comment by dmmiller** [comment]
I'm with @mkonicek, I think Spinner is a better name as well.  I think it's a pretty common name outside of Android for the component.

**Comment by janicduplessis** [comment]
Ok, let's change it to Spinner since you both agree on that :)

**Comment by satya164** [comment]
@dmmiller I also think that spinner is a better name, but this is going to cause confusion since another UI element exists with the same name.

How about calling it something like `ProgressIndicator` or `LoadingIndicator`?

File: Examples/UIExplorer/ActivityIndicatorExample.js [hunk 1]
```diff
@@ -15,41 +15,40 @@
...
+const {
+  ActivityIndicator,
   StyleSheet,
   View,
 } = ReactNative;
-var TimerMixin = require('react-timer-mixin');
+const TimerMixin = require('react-timer-mixin');
 
-var ToggleAnimatingActivityIndicator = React.createClass({
+console.log(ActivityIndicator);
```
Comment (dmmiller) [inline Examples/UIExplorer/ActivityIndicatorExample.js]: should be removed

**Comment by dmmiller** [comment]
There is a decent amount of cleanup mixed in here as well.  Would be nice to see that separate and then we can quickly approve that and then look at the meat of this PR.

**Comment by janicduplessis** [comment]
@dmmiller Good point. I'll try to separate this in multiple commits and add notes on what I changed.

File: Libraries/Components/ActivityIndicator/ActivityIndicatorIOS.ios.js [hunk 2]
```diff
@@ -7,27 +7,18 @@
  * of patent rights can be found in the PATENTS file in the same directory.
  *
  * @providesModule ActivityIndicatorIOS
- * @flow
```
Comment (janicduplessis) [inline Libraries/Components/ActivityIndicator/ActivityIndicatorIOS.ios.js]: Had to disable flow for this file, passing props to ActivityIndicator with `{...this.props}` caused a bunch of these errors. Not sure if there's a better way to fix this but the component is deprecated and there's pretty much no code anymore there.
```
 62:       <ActivityIndicator
           ^ React element `ActivityIndicator`
 37:     animating: PropTypes.bool,
                    ^^^^^^^^^^^^^^ undefined. This type is incompatible with. See: Libraries/Components/ActivityIndicator/ActivityIndicator.js:37
 60:       animating: true,
                      ^^^^ boolean. See: Libraries/Components/ActivityIndicator/ActivityIndicator.js:60

```

File: Libraries/Components/ActivityIndicator/ActivityIndicator.js [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    animating: PropTypes.bool,
+    /**
+     * The foreground color of the spinner (default is gray).
+     */
+    color: ColorPropType,
+    /**
+     * Size of the indicator. Small has a height of 20, large has a height of 36.
+     * Other sizes can be obtained using a scale transform.
+     */
+    size: PropTypes.oneOf([
```
Comment (janicduplessis) [inline Libraries/Components/ActivityIndicator/ActivityIndicator.js]: I wasn't sure exactly what sizes we want to support here. For iOS there are only 2 sizes possible, the only way to get more is to use a scale transform. For Android there were a few more size presets and it's possible to get any size by passing `width` and `height` to `RCTActivityIndicator`.

I used the same sizes for Android that iOS uses which mean that some of the size presets from ProgressBarAndroid are different / missing. The advantage of this is that the spinners will have the same size on both platform and it is still possible make the spinners any size using a scale transform.

**Comment by janicduplessis** [comment]
@dmmiller Split up the cleanup of ActivityIndicatorIOS in d7ecdd5, then a40fddf has most of the new stuff. Also added a few notes about stuff I wasn't 100% sure.

**Comment by janicduplessis** [comment]
@satya164 ProgressIndicator and LoadingIndicator are good but why not just call it ActivityIndicator in this case since it's very similar names and ActivityIndicator is already familiar to iOS devs. I like Spinner but my only concern is that it could cause some confusion with the native android view. That's why I'm still a bit biased towards keeping ActivityIndicator.

**Comment by satya164** [comment]
@janicduplessis I think `ActivityIndicator` is very iOSy not immediately familiar to web or Android devs. I think `LoadingIndicator` or `ProgressIndicator` sound more generic and obvious.

**Comment by mkonicek** [comment]
>  I like Spinner but my only concern is that it could cause some confusion with the native android view. That's why I'm still a bit biased towards keeping ActivityIndicator.

I'm fine with both but still a little bit in favor of Spinner. Not very concerned with Android naming, they are very bad at naming things :) Just look at `ProgressBar` - we'd have a similar issue there - "but Android developers are used to referring to a spinning indicator as a Progress Bar, we cannot name an actual horizontal progress bar ProgressBar now" :)

> I think ActivityIndicator is very iOSy.

Only for people who know iOS probably?

>  LoadingIndicator or ProgressIndicator sound more generic and obvious.

Those names are fine too. ProgressIndicator is more generic maybe as the app is not necessarily "loading" something.

I think we had this exact same discussion about naming internally and it always ends up with bikeshedding and we make no decision ;) @janicduplessis it's your call :)

**Comment by ide** [comment]
"ProgressIndicator" for me implies a sense of forward progress from 0 to 100% since that's how most progress bars work. Spinner makes a lot of sense but might be ambiguous -- JSpinner in Swing (yes that Swing) means a drop-down menu: https://docs.oracle.com/javase/tutorial/uiswing/components/spinner.html

**Comment by janicduplessis** [comment]
Here's what I like about `ActivityIndicator`

1. It doesn't imply any particular visual for the indicator, spinner implies it spins, what if it doesn't spin on iOS 10? (might be playing the devil's advocate here :))

2. It doesn't imply any determinate progress like `ProgressX`.

3. We already have `ActivityIndicatorIOS` and this has pretty much the same API so the name should not be confusing to anyone that used the old `ActivityIndicatorIOS`.

Unless someone completely disagree I think I'll go with this.

**Comment by sahrens** [comment]
I agree with "ActivityIndicator" - the only arguments I've heard against it are that its "iOS-y" and that its not what is used on Android, both of which are basically irrelevant - if there isn't one shared name across platforms, we should look at the name *independent* of other platforms to pick something that makes the most intuitive sense.

On Apr 26, 2016, at 1:13 AM, Janic Duplessis <notifications@github.com<mailto:notifications@github.com>> wrote:


Here's what I like about ActivityIndicator

  1.  It doesn't imply any particular visual for the indicator, spinner implies it spins, what if it doesn't spin on iOS 10? (might be playing the devil's advocate here :))

  2.  It doesn't imply any determinate progress like ProgressX.

  3.  We already have ActivityIndicatorIOS and this has pretty much the same API so the name should not be confusing to anyone that used the old ActivityIndicatorIOS.

Unless someone completely disagree I think I'll go with this.

**Comment by facebook-github-bot-8** [comment]
@janicduplessis updated the pull request.

**Comment by mkonicek** [comment]
Cool, agree with the naming.

@dmmiller Can we merge this?

**Comment by facebook-github-bot-9** [comment]
@ericvicenti would you mind taking a look at this pull request? It's been a while since the last commit was reviewed.

**Comment by satya164** [comment]
ping @dmmiller :)

**Comment by mkonicek** [comment]
I pinged Dave on messenger.

**Comment by dmmiller** [comment]
Let's do it.  I think we will have some code to fix up internally that will now be deprecated but I can do it.

@facebook-github-bot shipit

**Comment by facebook-github-bot-4** [comment]
Thanks for importing. If you are an FB employee go to [Phabricator](https://our.intern.facebook.com/intern/opensource/github/pull_request/807871229318970/int_phab) to review.

**Comment by facebook-github-bot-6** [comment]
I tried to merge this pull request into the Facebook internal repo but some checks failed. To unblock yourself please check the following: Does this pull request pass all open source tests on GitHub? If not please fix those. Does the code still apply cleanly on top of GitHub master? If not can please rebase. In all other cases this means some internal test failed, for example a part of a fb app won't work with this pull request. I've added the Import Failed label to this pull request so it is easy for someone at fb to find the pull request and check what failed. If you don't see anyone comment in a few days feel free to comment mentioning one of the core contributors to the project so they get a notification.

**Comment by facebook-github-bot-2** [comment]
@janicduplessis updated the pull request.

**Comment by dmmiller** [comment]
Can you rebase please?  It's failing on importing and applying the patch.  Also I think some of the commits could actually be separate cleanup pull requests since this will all get squashed into a single commit when it goes in.

**Comment by dmmiller** [comment]
@facebook-github-bot shipit

Let's try again.  Just saw you updated while I was commenting. :)

**Comment by janicduplessis** [comment]
Yea there was a conflict, should be good now :)

**Comment by mkonicek** [comment]
@facebook-github-bot shipit

**Comment by janicduplessis** [comment]
@facebook-github-bot shipit

**Comment by facebook-github-bot-9** [comment]
Thanks for importing. If you are an FB employee go to [Phabricator](https://our.intern.facebook.com/intern/opensource/github/pull_request/807871229318970/int_phab) to review.

**Comment by kamek-pf** [comment]
I feel like we're losing flexibility on the Android side because of this.
The only way to have a bigger size than the 'large' preset is to use a scale transform, but a spinner rendered this way looks much worse than what we had with `ProgressBarAndroid` 
Here's an example with a scale of 4 just, for the sake of argument :
screen
It looks just like a stretched image.

**Comment by dmmiller** [comment]
@kamek-pf, to be clear, nothing stops you from using ProgressBarAndroid if you need more control.  Or you can potentially put up a PR that adds support for more sizes.

**Comment by kamek-pf** [comment]
I'd be happy to keep using ProgressBarAndroid, it throws deprecation warning though : 
`Circular indeterminate ProgressBarAndroid is deprecated. Use ActivityIndicator instead.`

**Comment by janicduplessis** [comment]
I'm fine with adding something to control the size better, maybe  allow passing a number to the size prop that works only on android (could even be possible to make it work on ios by calculating a scale transform). 

As for now you can always add the warning to console.ignoredYellowBox to get rid of it.

**Comment by kamek-pf** [comment]
Sounds good, I think having a single component with platform specific properties makes a lot of sense in cases like this, where APIs are similar but not exactly the same.

**Comment by Noitidart** [comment]
I was not able to pull of a circular determinate progress per material guidelines here:


https://material.io/guidelines/components/progress-activity.html#progress-activity-types-of-indicators



Does anyone know how to achieve the determinate circular?

**Pull request closed without merging**